Reference Restructuring (#446)
vitessio/website: content/en/docs/reference/programs/mysqlctl.md

---
title: mysqlctl
---

`mysqlctl` is a command-line tool used for starting `mysqld` binaries. It is responsible for bootstrapping tasks such as generating a configuration file for `mysqld` and ensuring that `mysql_upgrade` is run in the data directory when restoring from backup.

`mysqld_safe` will be also be utilized when present. This helps ensure that `mysqld` is automatically restarted after failures.

## Commands

### init [-wait_time=5m] [-init_db_sql_file=(default)]

Bootstraps a new `mysqld` instance. The MySQL version and flavor will be auto-detected, with a minimal configuration file applied. For example:

```bash
export VTDATAROOT=/tmp
mysqlctl \
 -alsologtostderr \
 -tablet_uid 101 \
 -mysql_port 12345 \
 init
```

### init_config

Bootstraps the configuration for a new `mysqld` instance. This command is the same as `init` except the `mysqld` server will not be started. For example:

```bash
export VTDATAROOT=/tmp
mysqlctl \
 -alsologtostderr \
 -tablet_uid 101 \
 -mysql_port 12345 \
 init_config
```

### reinit_config

> Regenerate new configuration files for an existing `mysqld` instance. This could be helpful to revert configuration changes, or to pick up changes made to the bundled config in newer Vitess versions:

Reviewer: ```suggestion
Regenerate new configuration files for an existing `mysqld` instance. This could be helpful to revert configuration changes, or to pick up changes made to the bundled config in newer Vitess versions. For example:
```

Reviewer: I think it's best to keep this consistent throughout.

```bash
export VTDATAROOT=/tmp
mysqlctl \
 -alsologtostderr \
 -tablet_uid 101 \
 -mysql_port 12345 \
 reinit_config
```

### teardown [-wait_time=5m] [-force]

Remove the data files for a previously shutdown `mysqld` instance. This is a destructive operation:

```bash
export VTDATAROOT=/tmp
mysqlctl -tablet_uid 101 -alsologtostderr teardown
```

### start [-wait_time=5m]

Resume an existing `mysqld` instance that was previously bootstrapped with `init` or `init_config`:

```bash
export VTDATAROOT=/tmp
mysqlctl -tablet_uid 101 -alsologtostderr start
```

### shutdown [-wait_time=5m]

Stop a `mysqld` instance that was previously started with `init` or `start`.

For large `mysqld` instances, you may need to extend the -`wait_time` as flushing dirty pages.

```bash
export VTDATAROOT=/tmp
mysqlctl -tablet_uid 101 -alsologtostderr shutdown
```

## Options

The following global parameters apply to `mysqlctl`:

| Name | Type | Definition |
| :-------- | :--------- | :--------- |

> | alsologtostderr | boolean | log to standard error as well as files |

Reviewer: I think this table would be easier to read if the descriptions were consistently capitalized. I think capitalizing the first letter of each description is the way to go.

Reviewer: Also, capitalizing the first letter of each new sentence within the description.

Reviewer: It would be clearer to read if we took out the Latin abbreviations 'e.g.', 'i.e.', etc. Instead: 'For example,' 'that is.'

Author: These are autogenerated these from the program's output :-) We've fixed the easier ones in https://github.com/vitessio/vitess/pull/6261 , but for capitalization may defer to a future update. Thx for the detailed feedback.

| db_dba_password | string | db dba password |
| db_dba_use_ssl | boolean | Set this flag to false to make the dba connection to not use ssl (default true) |
| db_dba_user | string | db dba user userKey (default "vt_dba") |
| db_flags | uint | Flag values as defined by MySQL. |
| db_flavor | string | Flavor overrid. Valid value is FilePos. |
| db_host | string | The host name for the tcp connection. |
| db_port | int | tcp port |
| db_server_name | string | server name of the DB we are connecting to. |
| db_socket | string | The unix socket to connect on. If this is specified, host and port will not be used. |
| db_ssl_ca | string | connection ssl ca |
| db_ssl_ca_path | string | connection ssl ca path |
| db_ssl_cert | string | connection ssl certificate |
| db_ssl_key | string | connection ssl key |
| dba_idle_timeout | duration | Idle timeout for dba connections (default 1m0s) |
| dba_pool_size | int | Size of the connection pool for dba connections (default 20) |
| disable_active_reparents | boolean |  if set, do not allow active reparents. Use this to protect a cluster using external reparents. |

> | emit_stats | boolean | true iff we should emit stats to push-based monitoring/stats backends |

Reviewer: This description is hard to understand: who is 'we'? Probably it could be stated more clearly as something like, "If set, emit stats to push-based monitoring and stats backends."

| grpc_auth_mode | string | Which auth plugin implementation to use (eg: static) |
| grpc_auth_mtls_allowed_substrings | string | List of substrings of at least one of the client certificate names (separated by colon). |
| grpc_auth_static_client_creds | string | when using grpc_static_auth in the server, this file provides the credentials to use to authenticate with server |
| grpc_auth_static_password_file | string | JSON File to read the users/passwords from. |
| grpc_ca | string | ca to use, requires TLS, and enforces client cert check |
| grpc_cert | string | certificate to use, requires grpc_key, enables TLS |

> | grpc_compression | string | how to compress gRPC, default: nothing, supported: snappy |

Reviewer: ```suggestion
| grpc_compression | string | Which protocol to use for compressing gRPC. Default: nothing. Supported: snappy |
```

Reviewer: I think that we should stick with one punctuation pattern throughout, with capitalization and periods after each sentence in the description.

Reviewer 2: I wonder if we should even document this flag.

| grpc_enable_tracing | boolean | Enable GRPC tracing |
| grpc_initial_conn_window_size | int | grpc initial connection window size |
| grpc_initial_window_size | int | grpc initial window size |

> | grpc_keepalive_time | duration | After a duration of this time if the client doesn't see any activity it pings the server to see if the transport is still alive. (default 10s) |

Reviewer: ```suggestion
| grpc_keepalive_time | duration | After a duration of this time, if the client doesn't see any activity, it pings the server to see if the transport is still alive. (default 10s) |
```

| grpc_keepalive_timeout | duration | After having pinged for keepalive check, the client waits for a duration of Timeout and if no activity is seen even after that the connection is closed. (default 10s) |
| grpc_key | string | key to use, requires grpc_cert, enables TLS |
| grpc_max_connection_age | duration | Maximum age of a client connection before GoAway is sent. (default 2562047h47m16.854775807s) |
| grpc_max_connection_age_grace | duration | Additional grace period after grpc_max_connection_age, after which connections are forcibly closed. (default 2562047h47m16.854775807s) |
| grpc_max_message_size | int | Maximum allowed RPC message size. Larger messages will be rejected by gRPC with the error 'exceeding the max size'. (default 16777216) |
| grpc_port | int | Port to listen on for gRPC calls |
| grpc_prometheus | boolean | Enable gRPC monitoring with Prometheus |
| grpc_server_initial_conn_window_size | int | grpc server initial connection window size |
| grpc_server_initial_window_size | int | grpc server initial window size |

> | grpc_server_keepalive_enforcement_policy_min_time | duration | grpc server minimum keepalive time (default 5m0s) |

Reviewer: ```suggestion
| grpc_server_keepalive_enforcement_policy_min_time | duration | gRPC server minimum keepalive time (default 5m0s) |
```

| log_backtrace_at | value | when logging hits line file:N, emit a stack trace |
| log_dir | string | If non-empty, write log files in this directory |
| log_err_stacks | boolean | log stack traces for errors |
| log_rotate_max_size | uint | size in bytes at which logs are rotated (glog.MaxSize) (default 1887436800) |
| logtostderr | boolean | log to standard error instead of files |
| master_connect_retry | duration | how long to wait in between slave -> connection attempts. Only precise to the second. (default 10s) |
| mem-profile-rate | int | profile every n bytes allocated (default 524288) |
| mutex-profile-fraction | int | profile every n mutex contention events (see runtime.SetMutexProfileFraction) |
| mysql_auth_server_static_file | string | JSON File to read the users/passwords from. |
| mysql_auth_server_static_string | string | JSON representation of the users/passwords config. |
| mysql_auth_static_reload_interval | duration | Ticker to reload credentials |
| mysql_clientcert_auth_method | string | client-side authentication method to use. Supported values: mysql_clear_password, dialog. (default "mysql_clear_password") |
| mysql_port | int | mysql port (default 3306) |

> | mysql_server_flush_delay | duration | Delay after which buffered response will flushed to client. (default 100ms) |

Reviewer: ```suggestion
| mysql_server_flush_delay | duration | Delay after which buffered response flush to client. (default 100ms) |
```

Reviewer 2: "will be flushed". We should change the code to be consistent with this too.
